client/json-rpc/src/client.rs

```diff
@@ -143,54 +170,76 @@ impl JsonRpcAsyncClient {
     pub async fn get_accounts(
         &self,
         accounts: &[AccountAddress],
-    ) -> Result<Vec<Option<AccountView>>> {
+    ) -> Result<Vec<Option<AccountView>>, JsonRpcAsyncClientError> {
         let mut batch = JsonRpcBatch::new();
         for account in accounts {
             batch.add_get_account_request(*account);
         }
         let exec_results = self.execute(batch).await?;
         let mut results = vec![];
         for exec_result in exec_results {
-            let exec_result = exec_result?;
+            let exec_result =
+                exec_result.map_err(|e| JsonRpcAsyncClientError::ExecutionError(e.to_string()))?;
             if let JsonRpcResponse::AccountResponse(r) = exec_result {
                 results.push(r);
             } else {
                 panic!("Unexpected response for get_accounts {:?}", exec_result)
             }
         }
-        ensure!(
-            results.len() == accounts.len(),
-            "Received unexpected number of JSON RPC responses ({}) for {} requests",
-            results.len(),
-            accounts.len()
-        );
+
+        if results.len() != accounts.len() {
+            return Err(JsonRpcAsyncClientError::ExecutionError(format!(
+                "Received unexpected number of JSON RPC responses ({}) for {} requests",
+                results.len(),
+                accounts.len()
+            )));
+        }
+
         Ok(results)
     }
 
-    pub async fn submit_transaction(&self, txn: SignedTransaction) -> Result<()> {
+    pub async fn submit_transaction(
+        &self,
+        txn: SignedTransaction,
+    ) -> Result<(), JsonRpcAsyncClientError> {
         let mut batch = JsonRpcBatch::new();
-        batch.add_submit_request(txn)?;
+        batch
+            .add_submit_request(txn)
+            .map_err(|e| JsonRpcAsyncClientError::ExecutionError(e.to_string()))?;
         let mut exec_result = self.execute(batch).await?;
         assert!(exec_result.len() == 1);
-        exec_result.remove(0).map(|_| ())
+        exec_result
+            .remove(0)
+            .map(|_| ())
+            .map_err(|e| JsonRpcAsyncClientError::ExecutionError(e.to_string()))
     }
 
-    pub async fn execute(&self, batch: JsonRpcBatch) -> Result<Vec<Result<JsonRpcResponse>>> {
+    pub async fn execute(
+        &self,
+        batch: JsonRpcBatch,
+    ) -> Result<Vec<Result<JsonRpcResponse>>, JsonRpcAsyncClientError> {
         let requests = batch.json_request();
         let resp = self
             .client
             .post(&self.address)
             .json(&requests)
             .send()
-            .await?;
+            .await
+            .map_err(JsonRpcAsyncClientError::SubmissionError)?;
 
-        ensure!(
-            resp.status() == 200,
-            format!("Http error code {}", resp.status())
-        );
+        if resp.status() != 200 {
```

Comment:
I suspect that the specific type of http error could determine whether to retry or not, but I don't know the details. Should it be a TODO?

Reply:
If the return status is not in 200 range, it will already error out when we are sending the request.

Reply:
Huh, so you suspect this check is kind of redundant?